File: kvx/Models/RelayReminder.swift
//
//  RelayReminder.swift
//  kvx
//
//  Created by Developer Agent on 06/09/2026.
//

import Foundation

struct RelayReminder: Identifiable, Equatable, Hashable, Codable {
    let id: String
    let deviceId: String
    let relayIndex: Int
    let startTime: Date
    let duration: TimeInterval  // seconds
    let repeatType: ReminderRepeatType
    let isActive: Bool

    init(
        id: String = UUID().uuidString,
        deviceId: String,
        relayIndex: Int,
        startTime: Date,
        duration: TimeInterval,
        repeatType: ReminderRepeatType,
        isActive: Bool = true
    ) {
        precondition(relayIndex >= 0, "relayIndex must be >= 0")
        precondition(duration > 0, "duration must be positive")

        self.id = id
        self.deviceId = deviceId
        self.relayIndex = relayIndex
        self.startTime = startTime
        self.duration = duration
        self.repeatType = repeatType
        self.isActive = isActive
    }

    // Equatable based on id
    static func == (lhs: RelayReminder, rhs: RelayReminder) -> Bool {
        lhs.id == rhs.id
    }

    // Hashable based on id
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

File: kvx_flutter/lib/domain/entities/relay_reminder.dart
import 'reminder_repeat_type.dart';

class RelayReminder {
  final String id;
  final String deviceId;
  final int relayIndex;
  final DateTime startTime;
  final Duration duration;
  final ReminderRepeatType repeatType;
  final bool isActive;

  const RelayReminder({
    required this.id,
    required this.deviceId,
    required this.relayIndex,
    required this.startTime,
    required this.duration,
    required this.repeatType,
    this.isActive = true,
  }) : assert(relayIndex >= 0, 'relayIndex must be >= 0');

  RelayReminder copyWith({
    String? id,
    String? deviceId,
    int? relayIndex,
    DateTime? startTime,
    Duration? duration,
    ReminderRepeatType? repeatType,
    bool? isActive,
  }) {
    if (duration != null && duration <= Duration.zero) {
      throw ArgumentError('Duration must be positive');
    }

    return RelayReminder(
      id: id ?? this.id,
      deviceId: deviceId ?? this.deviceId,
      relayIndex: relayIndex ?? this.relayIndex,
      startTime: startTime ?? this.startTime,
      duration: duration ?? this.duration,
      repeatType: repeatType ?? this.repeatType,
      isActive: isActive ?? this.isActive,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RelayReminder &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}
